### PR TITLE
Added capsule geometry

### PR DESCRIPTION
**Pull request by ahcorde**
Added capsule geometry. Related with this issue https://github.com/ignitionrobotics/ign-rendering/issues/197

 - Ogre
capsule_ogre

 - Ogre2
capsule_ogre2

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

**Comment by ahcorde** [comment]
Depends on https://github.com/ignitionrobotics/ign-common/pull/155

**Comment by chapulina** [comment]
Same question as for the ellipsoid. Do we need a `Capsule` class? Maybe this is better, I'm just trying to understanding the rationale.

**Review by chapulina — CHANGES_REQUESTED** [review]
> Do we need a Capsule class?

Sorry for backtracking, but after the conversation in https://github.com/ignitionrobotics/ign-common/pull/165/, I think the capsule will need special treatment here.

Unlike the other simple shapes (box, sphere, cylinder, ellipsoid, or even cone in the future), a capsule can't be scaled to an arbitrary length + radius through the `SetLocalScale` function. Instead, `ign-rendering` will need to offer API that lets users change those 2 dimensions specifically.

**Comment by ahcorde** [comment]
API - https://github.com/ignitionrobotics/ign-rendering/blob/70554665bdcacff36b27a39f18c57ebba912c3ad/include/ignition/rendering/Capsule.hh

File: include/ignition/rendering/Capsule.hh [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      /// \brief Destructor
+      public: virtual ~Capsule() { }
+
+      public: virtual void SetRadius(const double _radius) = 0;
+
+      public: virtual void SetLength(const double _length) = 0;
+
+      public: virtual double Radius() = 0;
+
+      public: virtual double Length() = 0;
```
Comment (iche033) [inline include/ignition/rendering/Capsule.hh:46]: add doxygen comments for these functions

File: examples/gazebo_scene_viewer/SceneManager.cc [hunk 2]
```diff
@@ -1465,6 +1465,19 @@ void SubSceneManager::ProcessCone(
   _parent->AddGeometry(cone);
 }
 
+//////////////////////////////////////////////////
+void SubSceneManager::ProcessCapsule(
```
Comment (iche033) [inline examples/gazebo_scene_viewer/SceneManager.cc:1469]: is this function used?

File: include/ignition/rendering/base/BaseCapsule.hh [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+#include "ignition/rendering/Capsule.hh"
+#include "ignition/rendering/base/BaseObject.hh"
+
+namespace ignition
+{
+  namespace rendering
+  {
+    inline namespace IGNITION_RENDERING_VERSION_NAMESPACE {
+
+    /// \brief Base implementation of a Capsule Visual
```
Comment (iche033) [inline include/ignition/rendering/base/BaseCapsule.hh:32]: `..  a Capsule Geometry`

File: include/ignition/rendering/Capsule.hh [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    /// on the type specified.
+    class IGNITION_RENDERING_VISIBLE Capsule :
+      public virtual Geometry
+    {
+      /// \brief Destructor
+      public: virtual ~Capsule() { }
+
+      public: virtual void SetRadius(const double _radius) = 0;
+
+      public: virtual void SetLength(const double _length) = 0;
```
Comment (iche033) [inline include/ignition/rendering/Capsule.hh:42]: we started leaving out `const` for primitive type arguments

File: ogre2/include/ignition/rendering/ogre2/Ogre2Capsule.hh [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#ifndef IGNITION_RENDERING_OGRE2_OGRECAPSULE_HH_
+#define IGNITION_RENDERING_OGRE2_OGRECAPSULE_HH_
```
Comment (iche033) [inline ogre2/include/ignition/rendering/ogre2/Ogre2Capsule.hh:19]: IGNITION_RENDERING_OGRE2_OGRE**2**CAPSULE_HH_

File: ogre2/include/ignition/rendering/ogre2/Ogre2Capsule.hh [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        /// \param[in] _material Ogre material.
+        protected: virtual void SetMaterialImpl(Ogre2MaterialPtr _material);
+
+        /// \brief Create the Capsule Visual in ogre
+        private: void Create();
+
+        /// \brief Capsule should only be created by scene.
+        private: friend class Ogre2Scene;
+
+        private: void calculateRing(
```
Comment (iche033) [inline ogre2/include/ignition/rendering/ogre2/Ogre2Capsule.hh:79]: add doxygen comment, and also capitalize `C`, ie. `CalculateRing`

File: ogre2/include/ignition/rendering/ogre2/Ogre2Capsule.hh [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        public: virtual MaterialPtr Material() const;
+
+        // Documentation inherited.
+        public: virtual void SetMaterial(MaterialPtr _material, bool _unique);
+
+        /// \brief Set material to capsule geometry.
+        /// \param[in] _material Ogre material.
+        protected: virtual void SetMaterialImpl(Ogre2MaterialPtr _material);
+
+        /// \brief Create the Capsule Visual in ogre
```
Comment (iche033) [inline ogre2/include/ignition/rendering/ogre2/Ogre2Capsule.hh:73]: `... Create the Capsule geometry ...`

**Review by iche033 — COMMENTED** [review]
did a quick first pass. The Capsule geometry class looks good to me!

File: ogre2/src/Ogre2Capsule.cc [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+  // Set material for the underlying dynamic renderable
+  this->dataPtr->ogreMesh->SetMaterial(_material, false);
+  this->SetMaterialImpl(derived);
+}
+
+//////////////////////////////////////////////////
+void Ogre2Capsule::SetMaterialImpl(Ogre2MaterialPtr _material)
+{
+  Ogre::MaterialPtr ogreMaterial = _material->Material();
```
Comment (chapulina) [inline ogre2/src/Ogre2Capsule.cc:162]: Did you mean to use the `ogreMaterial` variable?

File: src/Capsule_TEST.cc [hunk 9]
```diff
@@ -0,0 +1,100 @@
+/* * Copyright (C) 2021 Open Source Robotics Foundation
```
Comment (chapulina) [inline src/Capsule_TEST.cc:1]:
```suggestion
/*
 * Copyright (C) 2021 Open Source Robotics Foundation
```

File: ogre2/src/Ogre2Camera.cc [hunk 10]
```diff
@@ -220,7 +220,7 @@ VisualPtr Ogre2Camera::VisualAt(const ignition::math::Vector2i &_mousePos)
     }
   }
 
-  float ratio = screenScalingFactor();
+  float ratio = 1;//screenScalingFactor();
```
Comment (chapulina) [inline ogre2/src/Ogre2Camera.cc:223]: Is this change related to capsules?

File: ogre2/include/ignition/rendering/ogre2/Ogre2Capsule.hh [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        public: virtual ~Ogre2Capsule();
+
+        // Documentation inherited.
+        public: virtual void Init() override;
+
+        // Documentation inherited.
+        public: virtual void Destroy() override;
+
+        // Documentation inherited.
+        public: virtual Ogre::MovableObject *OgreObject() const;
```
Comment (chapulina) [inline ogre2/include/ignition/rendering/ogre2/Ogre2Capsule.hh:58]: Same thing here about using `override`

File: ogre/src/OgreCapsule.cc [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+  }
+
+  this->SetMaterialImpl(derived);
+}
+
+//////////////////////////////////////////////////
+void OgreCapsule::SetMaterialImpl(OgreMaterialPtr _material)
+{
+  std::string materialName = _material->Name();
+  Ogre::MaterialPtr ogreMaterial = _material->Material();
```
Comment (chapulina) [inline ogre/src/OgreCapsule.cc:166]: `materialName` and `ogreMaterial` aren't used, did you mean to use them?

File: ogre2/src/Ogre2Capsule.cc [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+  if (!derived)
+  {
+    ignerr << "Cannot assign material created by another render-engine"
+        << std::endl;
+
+    return;
+  }
+
+  // Set material for the underlying dynamic renderable
+  this->dataPtr->ogreMesh->SetMaterial(_material, false);
```
Comment (chapulina) [inline ogre2/src/Ogre2Capsule.cc:155]: Should this `ogreMesh->SetMaterial` call be in `SetMaterialImpl`? Otherwise if another function calls that it won't have any effect.

In fact, I'm not sure what's the advantage of having a separate `SetMaterialImpl` function, it looks to me that the entire logic could be inside `SetMaterial`.

File: ogre2/src/Ogre2Capsule.cc [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+#include "ignition/rendering/ogre2/Ogre2Scene.hh"
+#include "ignition/rendering/ogre2/Ogre2Mesh.hh"
+#include "ignition/rendering/ogre2/Ogre2Visual.hh"
+
+#include <ignition/math/Vector3.hh>
+
+class ignition::rendering::Ogre2CapsulePrivate
+{
+  /// \brief Capsule materal
+  public: Ogre2MaterialPtr material;
```
Comment (chapulina) [inline ogre2/src/Ogre2Capsule.cc:34]:
```suggestion
  public: Ogre2MaterialPtr material{nullptr};
```

**Comment by ahcorde** [comment]
I'm getting this error 

```
simple_demo: /var/lib/jenkins/workspace/ogre-2.1-debbuilder/repo/OgreMain/src/OgreHlmsDatablock.cpp:170: virtual Ogre::HlmsDatablock::~HlmsDatablock(): Assertion `mLinkedRenderables.empty() && "This Datablock is still being used by some Renderables." " Change their Datablocks before destroying this."' failed.
Abortado (`core' generado)
```

I'm not really sure why

**Comment by iche033** [comment]
> I'm getting this error

When do you get that error?

File: include/ignition/rendering/Capsule.hh [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+      public: virtual void SetRadius(double _radius) = 0;
+
+      /// \brief Set the length of the capsule
+      public: virtual void SetLength(double _length) = 0;
+
+      /// \brief Get the radius of the capsule
+      public: virtual double Radius() = 0;
+
+      /// \brief Get the length of the capsule
+      public: virtual double Length() = 0;
```
Comment (iche033) [inline include/ignition/rendering/Capsule.hh:50]: `Radius()` and `Length()` can be const functions

File: examples/simple_demo/Main.cc [hunk 16]
```diff
@@ -137,6 +137,17 @@ void buildScene(ScenePtr _scene)
   white->SetRenderOrder(0);
 //! [white material]
 
+  VisualPtr capsule_visual = _scene->CreateVisual();
```
Comment (iche033) [inline examples/simple_demo/Main.cc:140]: camelCase

**Comment by ahcorde** [comment]
> When do you get that error?

I fixed the error in one of the commits

**Review by chapulina — CHANGES_REQUESTED** [review]
I have some suggestions for this PR on #276

**Review by chapulina — APPROVED** [review]
:pill: :tada:

File: ogre2/src/Ogre2Capsule.cc [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+////////////////////////////////////////////////
+void Ogre2Capsule::Create()
+{
+  common::MeshManager *meshMgr = common::MeshManager::Instance();
+  std::string capsuleMeshName = this->Name() + "_capsule_mesh"
+    + "_" + std::to_string(this->radius)
+    + "_" + std::to_string(this->length);
+  if (!meshMgr->HasMesh(capsuleMeshName))
+  {
+    meshMgr->CreateCapsule(capsuleMeshName, this->radius, this->length, 12, 32);
```
[the remark this answers is not in the document]
Comment (chapulina) [reply ogre2/src/Ogre2Capsule.cc]: Made them the same on #276
